jobs/dtd-metrics/src/services/metrics-producer.service.ts

```diff
@@ -0,0 +1,92 @@
+import { ReadModelClient } from '@interop-be-reports/commons'
+import { MetricsOutput } from '../models/metrics.model.js'
+import { GlobalStoreService } from './global-store.service.js'
+import { writeFileSync } from 'fs'
+import { log, timer } from '../utils/helpers.utils.js'
+
+export type MetricFactoryFn<TMetricKey extends keyof MetricsOutput> = (
+  readModel: ReadModelClient,
+  globalStore: GlobalStoreService
+) => Promise<MetricsOutput[TMetricKey]> | MetricsOutput[TMetricKey]
+
+type MetricObj<TMetricKey extends keyof MetricsOutput> = {
+  name: TMetricKey
+  factoryFn: MetricFactoryFn<TMetricKey>
+}
+
+type ProduceOutputOptions = {
+  produceJSON?: boolean
+  filter?: string
+}
+
+export class MetricsProducerService {
+  private metrics: Array<MetricObj<keyof MetricsOutput>> = []
+
+  constructor(
+    private readModel: ReadModelClient,
+    private globalStore: GlobalStoreService
+  ) {}
+
+  /**
+   * Adds a metric to the metrics list.
+   */
+  public addMetric<TMetricKey extends keyof MetricsOutput, TMetricFactoryFn extends MetricFactoryFn<TMetricKey>>(
+    metricName: TMetricKey,
+    metricFactory: TMetricFactoryFn
+  ): MetricsProducerService {
+    this.metrics.push({
+      name: metricName,
+      factoryFn: this.wrapFactoryFnWithLogs(metricName, metricFactory),
+    })
+
+    return this
+  }
+
+  /**
+   * Executes all the metrics sequentially and returns a MetricsOutput object.
+   * If a filter is specified, only the metrics whose name includes the filter will be executed.
+   * If produceJSON is true, the output will be written to a file named `dtd-metrics.json`.
+   */
+  public async produceOutput({
+    produceJSON,
+    filter,
+  }: ProduceOutputOptions): Promise<MetricsOutput | Partial<MetricsOutput>> {
+    const metricsOutput: Record<string, unknown> = {}
+
+    for (const metricObj of this.metrics) {
+      // If a filter is specified, skip the metric if its name does not include the filter
+      if (filter && !metricObj.name.includes(filter)) continue
```

Comment:
not sure about the purpose of this filter, is it only for development?
if not, shouldn't be better if we entirely avoid the calculation if a metric is not required?

Reply:
It is only for development use. It helps a lot when I need to concentrate on a specific metric. Do you think I should make this clear?

Reply:
if it's just for development it's fine, not my intention to overcomplicate the logic if not required 😁